src/auth/index.js

```diff
@@ -129,149 +131,177 @@ export default class Auth {
+                { refreshOnFailure: false, useRefreshToken: true })
+    )
+    .then(response => response.json())
+    .then(json => this.set(json))
+    .catch(e => {
+      return this.stopImpersonation().then(() => {
+        throw e; // rethrow
       });
+    });
   }
 
-  async startImpersonation(client, userId) {
-    if (!(await this.authedId())) {
-      return Promise.reject(new StitchError('Must auth first'));
-    }
-
-    if (await this.isImpersonatingUser()) {
-      return Promise.reject(new StitchError('Already impersonating a user'));
-    }
-
-    await this.storage.set(authCommon.IMPERSONATION_ACTIVE_KEY, 'true');
-    await this.storage.set(authCommon.IMPERSONATION_USER_KEY, userId);
+  startImpersonation(client, userId) {
+    return this.authedId().then((authedId) => {
+      if (!authedId) {
+        Promise.reject(new StitchError('Must auth first'))
+      }
+      return this.isImpersonatingUser();
+    }).then((isImpersonatingUser) => {
+      if (isImpersonatingUser) {
+        return Promise.reject(new StitchError('Already impersonating a user'));
+      }
 
-    let realUserAuth = JSON.parse(await this.storage.get(authCommon.USER_AUTH_KEY));
-    await this.storage.set(authCommon.IMPERSONATION_REAL_USER_AUTH_KEY, JSON.stringify(realUserAuth));
-    return this.refreshImpersonation(client);
+      return this.storage.set(authCommon.IMPERSONATION_ACTIVE_KEY, 'true');
+    }).then(() =>
+      this.storage.set(authCommon.IMPERSONATION_USER_KEY, userId)
+    ).then(() => 
+      this.storage.get(authCommon.USER_AUTH_KEY)
+    ).then((userAuth) => {
+      let realUserAuth = JSON.parse(userAuth);
+      return this.storage.set(authCommon.IMPERSONATION_REAL_USER_AUTH_KEY, JSON.stringify(realUserAuth));
+    }).then(() => this.refreshImpersonation(client));
   }
 
-  async stopImpersonation() {
-    if (!(await this.isImpersonatingUser())) {
-      throw new StitchError('Not impersonating a user');
-    }
+  stopImpersonation() {
+    return this.isImpersonatingUser().then((isImpersonatingUser) => {
+      if (!isImpersonatingUser) {
+        throw new StitchError('Not impersonating a user');
+      }
 
-    let realUserAuth = JSON.parse(await this.storage.get(authCommon.IMPERSONATION_REAL_USER_AUTH_KEY));
-    await this.set(realUserAuth);
-    await this.clearImpersonation();
+      return this.storage.get(authCommon.IMPERSONATION_REAL_USER_AUTH_KEY);
+    }).then((userAuth) => {
+      let realUserAuth = JSON.parse(userAuth);
+      return this.set(realUserAuth);
+    }).then(() => this.clearImpersonation());
   }
 
-  async clearImpersonation() {
-    await this.storage.remove(authCommon.IMPERSONATION_ACTIVE_KEY);
-    await this.storage.remove(authCommon.IMPERSONATION_USER_KEY);
-    await this.storage.remove(authCommon.IMPERSONATION_REAL_USER_AUTH_KEY);
+  clearImpersonation() {
+    return this.storage.remove(authCommon.IMPERSONATION_ACTIVE_KEY).then(() => 
```

Comment:
maybe `Promise.all` would work nicely here

Reply:
What's the advantage of `Promise.all` here?

Reply:
It accomplishes the same thing, but the resulting code is smaller/less noisy:
```
return Promise.all([
    this.storage.remove(authCommon.IMPERSONATION_ACTIVE_KEY),
    this.storage.remove(authCommon. IMPERSONATION_REAL_USER_AUTH_KEY),
    this.storage.remove(authCommon. IMPERSONATION_USER_KEY),
])
```

Also, we don't actually care about enforcing a strict ordering of how these keys get deleted, so chaining them together with `.then` isn't necessary (and maybe misleading to someone reading the code).